zhaquirks/legrand/remote_switch.py

```diff
@@ -0,0 +1,218 @@
+
+    def handle_cluster_request(
+        self,
+        hdr: foundation.ZCLHeader,
+        args: List[Any],
+        *,
+        dst_addressing: Optional[
+            Union[t.Addressing.Group, t.Addressing.IEEE, t.Addressing.NWK]
+        ] = None,
+    ):
+        """handle_cluster_request."""
+
+        if hdr.command_id in (0x0000, 0x0001, 0x0002):
+            # Send default response
+            if not hdr.frame_control.disable_default_response:
+                self.send_default_rsp(hdr, status=foundation.Status.SUCCESS)
+
+            self.endpoint.device.voltage_bus.listener_event(VOLTAGE_REPORTED)
+        else:
+            return super().handle_cluster_request(
+                hdr, args, dst_addressing=dst_addressing
+            )
+
+
+class LegrandRemoteSwitchPowerConfiguration(PowerConfigurationCluster):
+    """Legrand Wireless Remote Switch On/Off custom cluster."""
+
+    cluster_id = 0x0001
+    name = "PowerConfiguration"
+
+    _CONSTANT_ATTRIBUTES = {
+        0x0031: 10,  # battery_size = CR2032
+        0x0033: 1,  # battery_quantity
+        0x0034: 30,  # battery_rated_voltage = 3000 mV
+    }
+
+    def __init__(self, *args, **kwargs):
+        """Init."""
+
+        super().__init__(*args, **kwargs)
+        self.endpoint.device.voltage_bus.add_listener(self)
+
+    async def _configure_reporting(self, *args, **kwargs):  # pylint: disable=W0221
+        """Prevent remote configure reporting."""
+
+        return (foundation.ConfigureReportingResponse.deserialize(b"\x00")[0],)
+
+    def voltage_reported(self):
+        """Voltage reported."""
+
+        self.create_catching_task(
+            super().read_attributes(
+                [
+                    self.BATTERY_VOLTAGE_ATTR,
+                ],
+                manufacturer=0x1021,
+            )
+        )
+
+
+class LegrandRemoteSwitchPollControl(CustomCluster, PollControl):
+    """Legrand Wireless Remote Switch PollControl cluster.
+
+    Read PowerConfiguration attributes during check-in.
+    """
+
+    cluster_id = 0x0020
+    name = "PollControl"
+
+    def __init__(self, *args, **kwargs):
+        """Init."""
+
+        self._current_state = {}
+        super().__init__(*args, **kwargs)
+
+    def handle_cluster_request(
+        self,
+        hdr: foundation.ZCLHeader,
+        args: List[Any],
+        *,
+        dst_addressing: Optional[
+            Union[t.Addressing.Group, t.Addressing.IEEE, t.Addressing.NWK]
+        ] = None,
+    ):
+        """handle_cluster_request."""
+
+        if hdr.command_id == 0x0000:
+            self.endpoint.device.voltage_bus.listener_event(VOLTAGE_REPORTED)
+            return super().handle_cluster_request(
+                hdr, args, dst_addressing=dst_addressing
+            )
+
+
+class RemoteSwitch(CustomDevice):
+    """Wireless Remote switch 067773 (NLT type device)."""
+
+    def __init__(self, *args, **kwargs):
+        """Init."""
+        self.voltage_bus = Bus()
```

Comment:
You don’t need bus you can directly update cluster attributes now like this: https://github.com/zigpy/zha-device-handlers/blob/af69611ef39d516f7637f1ec2cd8c5a1c7f05692/zhaquirks/xiaomi/aqara/motion_ac01.py#L69

Reply:
I'm not sure I fully understand, as I'm not an expert neither in Zigbee nor in Python :( My idea is actually very simple. I just want to read voltage attribute every time on/off button is pressed and on every device check-ins. 

So, should I completely remove bus in favor of this new approach? Would it be possible to call `read_attributes` as a catching task from OnOff and PollControl clusters?

Reply:
Battery devices sleep. Are you sure reporting doesn’t work? Battery values may not report for days at a time

Reply:
Reporting never worked for me. My quirk was actually inspired by this zigbee2mqtt converter:

https://github.com/Koenkk/zigbee-herdsman-converters/blob/96563772e92eb459bf52c10e241c3a725968bcab/devices/legrand.js#L119

Reply:
This is a stock device. Looking at that converter it doesn’t prevent reporting. It reads the initial value on join which we already do. You can remove all overridden clusters from this except power configuration. I’ll add new comments in the files

Reply:
Thanks a lot for your support.

I tried your approach. I have deleted everything but the `PowerConfigurationCluster`:

```
class LegrandRemoteSwitchPowerConfiguration(PowerConfigurationCluster):
    """Legrand Wireless Remote Switch On/Off custom cluster."""

    _CONSTANT_ATTRIBUTES = {
        0x0031: 10,  # battery_size = CR2032
        0x0033: 1, # battery_quantity
        0x0034: 30,  # battery_rated_voltage = 3000 mV
    }
    
    MIN_VOLTS = 2.5
    MAX_VOLTS = 3.0
```

I removed and rejoined the device. Indeed, at initialization I saw voltage readings:

```
2022-10-17 00:42:06.188 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x0001] Sending request header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=False, direction=<Direction.Server_to_Client: 0>, disable_default_response=0, reserved=0, *is_cluster=False, *is_general=True, *is_reply=False), tsn=60, command_id=<GeneralCommand.Read_Attributes: 0>, *direction=<Direction.Server_to_Client: 0>, *is_reply=False)
2022-10-17 00:42:06.189 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x0001] Sending request: Read_Attributes(attribute_ids=[32, 33])
2022-10-17 00:42:06.190 DEBUG (MainThread) [zigpy.device] [0x950b] Extending timeout for 0x3c request
2022-10-17 00:42:06.194 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x000f] Sending request header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=False, direction=<Direction.Server_to_Client: 0>, disable_default_response=0, reserved=0, *is_cluster=False, *is_general=True, *is_reply=False), tsn=62, command_id=<GeneralCommand.Read_Attributes: 0>, *direction=<Direction.Server_to_Client: 0>, *is_reply=False)
2022-10-17 00:42:06.195 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x000f] Sending request: Read_Attributes(attribute_ids=[85])
2022-10-17 00:42:06.196 DEBUG (MainThread) [zigpy.device] [0x950b] Extending timeout for 0x3e request
2022-10-17 00:42:06.743 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x0001] Received ZCL frame: b'\x18<\x01 \x00\x00 \x1b!\x00\x86'
2022-10-17 00:42:06.745 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x0001] Decoded ZCL frame header: ZCLHeader(frame_control=FrameControl(frame_type=<FrameType.GLOBAL_COMMAND: 0>, is_manufacturer_specific=0, direction=<Direction.Client_to_Server: 1>, disable_default_response=1, reserved=0, *is_cluster=False, *is_general=True, *is_reply=True), tsn=60, command_id=1, *direction=<Direction.Client_to_Server: 1>, *is_reply=True)
2022-10-17 00:42:06.749 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x0001] Decoded ZCL frame: LegrandRemoteSwitchPowerConfiguration:Read_Attributes_rsp(status_records=[ReadAttributeRecord(attrid=0x0020, status=<Status.SUCCESS: 0>, value=TypeValue(type=uint8_t, value=27)), ReadAttributeRecord(attrid=0x0021, status=<Status.UNSUPPORTED_ATTRIBUTE: 134>)])
2022-10-17 00:42:06.751 DEBUG (MainThread) [zigpy.zcl] [0x950B:1:0x0001] Voltage [RAW]:27 [Max]:3.0 [Min]:2.5, Battery Percent: 40.0
```

But it seems that it didn't read voltage at device network join / device announce? How to check this and to ensure this?

Reply:
What do you mean? On initial join this will happen. if you are resetting / repairing the device without removing it from HA first it won't.

Reply:
I understand, that the voltage is read on initial join. But after that voltage is never read/updated: 1) devices itself doesn't report voltage, and 2) ZHA doesn't read voltage explicitly. As a result, voltage is not updated in ZHA at all. 

zigbee2mqtt "manually" and explicitly read voltage attribute on "device announcement". In my initial patch, I read voltage attribute on every button press.

Reply:
This is not true. https://github.com/Koenkk/zigbee-herdsman-converters/blob/96563772e92eb459bf52c10e241c3a725968bcab/devices/legrand.js#L133

This line is setting up attribute reporting for power configuration. Where is this converter manually reading voltage?

Reply:
For this part: https://github.com/Koenkk/zigbee-herdsman-converters/blob/e123e5e632800f10f8003cb706384cc7c7b8d23c/devices/legrand.js#L11

This is called on device announcement but doesn’t get called on every single interaction like this quirk was doing. Again, battery reports are only sent if they need to be and these batteries can last up to 2 years in these devices. Are you sure that reporting isn’t working?

Reply:
We also re-read this on device announcements: https://github.com/home-assistant/core/blob/6b56336e52b512ef22906d824a4563a584a664cf/homeassistant/components/zha/core/channels/general.py#L500

this is called from async_initialize via device_became_available on device which is called from device_rejoined in gateway which is what is called after we are sent a device announcement